packages/expo-modules-autolinking/src/autolinking.ts

```diff
@@ -62,84 +62,203 @@ export async function findDefaultPathsAsync(cwd: string): Promise<string[]> {
   return paths;
 }
 
+/**
+ * Finds the real path to custom native modules directory.
+ * @returns undefined if custom modules dir not found or doesn't exist
+ */
+export async function resolveNativeModulesDirAsync(
+  nativeModulesDir: string | undefined,
+  cwd: string
+): Promise<string | undefined> {
+  // first try resolving the provided dir
+  if (nativeModulesDir) {
+    const nativeModulesDirPath = path.resolve(nativeModulesDir);
+    if (await fs.pathExists(nativeModulesDirPath)) {
+      return nativeModulesDirPath;
+    }
+  }
+
+  // if not found, try to find it relative to the package.json
+  const up = await findUp('package.json', { cwd });
+  if (!up) {
+    return undefined;
```

Comment:
I don't have very strong opinion on this, but I would prefer `null` instead. `undefined` was never intended to be used explicitly by devs — it may mean that 1) function doesn't return any value, 2) something is unset/unknown, 3) someone forgot to return a value in the edge case. When you use `null`, it's clear for others who call this function that `null` was returned intentionally.